doc/design/draft/subscription.md

```diff
@@ -0,0 +1,180 @@
+
+| Publisher | Timepoint | Sample | | Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|-|:---------:|:---------:|:------:|
+| P1        | 1         | A      | | P2        | 20        | F      |
+| P1        | 11        | B      | | P2        | 22        | G      |
+| P1        | 21        | C      | | P2        | 24        | H      |
+| P1        | 31        | D      | | P2        | 40        | I      |
+| P1        | 41        | E      | | P2        | 42        | J      |
+
+Let's assume the subscriber subscribes to the service.
+
+#### Subscription with _HistorySize_ 0
+
+The subscriber will not be provided with samples from the history.
+
+#### Subscription with _HistorySize_ 1
+
+The subscriber cache contains the one sample ordered after _DeliveryTime_.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+
+#### Subscription with _HistorySize_ 6
+
+The subscriber cache contains the six samples ordered after _DeliveryTime_.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+| P2        | 40        | I      |
+| P1        | 31        | D      |
+| P2        | 24        | H      |
+| P2        | 22        | G      |
+
+#### Subscription with _HistorySize_ 20
+
+The subscriber cache contains the ten samples ordered after _DeliveryTime_.
+It contains only ten since there are no more samples available on publisher side.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+| P2        | 40        | I      |
+| P1        | 31        | D      |
+| P2        | 24        | H      |
+| P2        | 22        | G      |
+| P1        | 21        | C      |
+| P2        | 20        | F      |
+| P1        | 11        | B      |
+| P1        | 1         | A      |
+
+### Subscribing with concurrent send
+
+Let's assume publisher P1 would like to deliver the sample K at timepoint
+52 while the subscriber requested subsription. Furthermore, the subscriber
+requests always a _HistorySize_ of 2.
+
+All the listed cases should behave identical if there is more then one
+concurrent publish in progress.
+
+#### Publisher before _DeliveryTime_
+
+The subscriber cache contains two samples.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+
+The sample K will be delivered after successful subscription.
+
+#### Publisher after _DeliveryTime_
+
+The subscriber cache contains two samples.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P1        | 51        | K      |
+| P2        | 42        | J      |
+
+The sample K will not be delivered by the publisher since the _DeliveryTime_
+is defined as exactly the timepoint after the sample is placed into the ChunkHistory
```

Comment:
There is no such timepoint **exactly** after some observable event happens. In the distributed setting something happens before something else without meaningful timepoints or the order of two events is indeterminate (and not observed). We can define timepoints for the time some event became observable, but care must be taken.

There is no consistent way to define the point after something happens, unless we have a global synchronized logical clock (then it may be define as the next tick).

Reply:
@MatthiasKillat If we put the sample into the ChunkHistory within a mutex and set the timestamp in that mutex scope as well we should have the exact timepoint. The global synchronized logical clock would be the incremented number in comment: https://github.com/eclipse-iceoryx/iceoryx/pull/780/files#r654320386

Reply:
Should work if you have a mutex like today or some other kind of atomic operations that define a "happened before"